src/contrib/mlir/backend/pass/affine_lowerer.cpp

```diff
@@ -678,6 +715,207 @@ namespace
         return matchSuccess();
     }
 
+    REWRITER(NGMatMulOp)
+    {
+        auto matmul = cast<NGMatMulOp>(op);
+        auto loc = matmul.getLoc();
+
+        // Retrieve/generate Values for operands and result.
+        ScopedContext scope(rewriter, loc);
+        Value* lhs = operands[0];
+        Value* rhs = operands[1];
+        Value* result = pass.buildOutputDefs(op, rewriter)[0];
+        NGRAPH_CHECK(lhs && rhs && result, "Unexpected null values in MatMulOp");
+
+        auto resultTy = result->getType().dyn_cast<MemRefType>();
+        auto resultShape = resultTy.getShape();
+        auto lhsTy = lhs->getType().dyn_cast<MemRefType>();
+        auto lhsShape = lhsTy.getShape();
+        auto rhsTy = rhs->getType().dyn_cast<MemRefType>();
+        auto rhsShape = rhsTy.getShape();
+        NGRAPH_CHECK(resultTy, "Unexpected non-memref result type");
+        NGRAPH_CHECK(lhsTy, "Unexpected non-memref LHS type");
+        NGRAPH_CHECK(rhsTy, "Unexpected non-memref RHS type");
+
+        Type elemTy = resultTy.getElementType();
+        NGRAPH_CHECK(elemTy == lhsTy.getElementType() && elemTy == rhsTy.getElementType(),
+                     "Types mismatch in MatMulOp");
+
+        NGRAPH_CHECK(lhsShape.size() == 2 && rhsShape.size() == 2 && resultShape.size() == 2,
+                     "MatMul operation is only supported for 2D tensors");
+
+        auto attrPtr = static_cast<gemmAttrs*>(malloc(sizeof(gemmAttrs)));
```

Comment:
This looks like a bug. One compile-time alloc and free one each callback return. I suggest just having a union of all structs you need as a static member of the pass (or call-back utils), and just pass a pointer to that.

Reply:
We may have multiple callbacks.  A union would not work because we need one struct for each callback. How about use a vector of void* to store those pointers? We call free in destrutor of MLIRCPURuntime.

Reply:
Oh right, you are populating the structs at compile-time. Sure a list of pointers would do.